### PR TITLE
integrity: scaffolding for integrity check

### PR DESCRIPTION
**Pull request by AlgebraicWolf**
This patch adds necessary scaffolding for performing integrity checks,
along with an appropriate instrumentation of operations with files
that must be scrutinized before working with them.

### CODE REVIEW

File: cli/integrity/repository.go [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+import "io"
+
+// Repository provides utilities for working with files and
+// ensuring that they were not compomised.
+type Repository interface {
+	// AddHash extends the list of supported hashes.
+	AddHash(h any)
+	// Add checks signature of provided hashes file, and adds files
+	// to the repository.
+	Add(path string) error
```
Comment (oleg-jukovec) [inline cli/integrity/repository.go:12]: It looks like this should be specified when constructing a repository, which means it should be in the implementation constructor, and not in the general interface.

File: cli/integrity/integrity.go [hunk 2]
```diff
@@ -6,6 +6,8 @@ import (
 	"github.com/spf13/pflag"
 )
 
+var FileRepository Repository = dummyRepository{}
```
Comment (oleg-jukovec) [inline cli/integrity/integrity.go:9]: A global object is not a good idea. It's hard to test, it's hard to maintain.

It's much better if it's tied to some common command context. Is it too hard to implement?

File: cli/cluster/file.go [hunk 3]
```diff
@@ -20,7 +23,13 @@ func NewFileCollector(path string) FileCollector {
...
 // Collect collects a configuration from a file located at a specified path.
 func (collector FileCollector) Collect() (*Config, error) {
-	data, err := os.ReadFile(collector.path)
+	file, err := integrity.FileRepository.Read(collector.path)
+	if err != nil {
+		return nil, fmt.Errorf("can't check integrity of %q: %w", collector.path, err)
+	}
+	defer file.Close()
+
+	data, err := io.ReadAll(file)
```
Comment (oleg-jukovec) [inline cli/cluster/file.go:32]: I would expect here different implementations in `integrity` module for config collectors/publishers.

This is not so obvious in the context of the file collector, but in other cases the algorithms are too different. As example, you could try to implement `FileDataPublisher` in that way.

File: cli/integrity/integrity.go (hunk 2 above)
Comment (AlgebraicWolf) [reply cli/integrity/integrity.go:9]: I don't think it would be _prohibitively_ hard, but there are certainly places where the command context is not passed, but we would like to use the integrity module.

Comment (oleg-jukovec) [reply cli/integrity/integrity.go:9]: What places?

Comment (oleg-jukovec) [reply cli/integrity/integrity.go:9]: > In case of script/cluster instances, this is also not immediately obvious, although I believe we can store a pointer to a repository inside the instance structure.

What the case?

File: cli/cluster/file.go (hunk 3 above)
Comment (AlgebraicWolf) [reply cli/cluster/file.go:32]: Why would `FileDataPublisher` be affected in any way by this?

Comment (AlgebraicWolf) [reply cli/cluster/file.go:32]: Thank you, I see what you've meant now!

File: cli/cluster/file.go
Comment (oleg-jukovec) [inline cli/cluster/file.go:1]: Please, revert changes in the file. I've implemented it in another one:
https://github.com/tarantool/tt/pull/714

File: cli/cmd/root.go [hunk 4]
```diff
@@ -199,13 +202,33 @@ func InitRoot() {
...
+		if err != nil {
+			log.Fatalf("can't get current dir: %w", err.Error())
+		}
+
+		configPath, _ := util.GetYamlFileName(
+			filepath.Join(currentDir, configure.ConfigName),
+			false,
+		)
+		// Remove the following line after debugging.
+		// log.SetLevel(log.DebugLevel)
```
Comment (oleg-jukovec) [inline cli/cmd/root.go:218]: Please, cleanup the debug comment.

File: cli/configure/configure.go [hunk 5]
```diff
@@ -195,8 +196,14 @@ func GetCliOpts(configurePath string) (*config.CliOpts, string, error) {
 	var cfg *config.CliOpts = GetDefaultCliOpts()
 	// Config could not be processed.
 	configPath, err := util.GetYamlFileName(configurePath, true)
+	// Vefore loading configure file, we'll initialize integrity checking.
```
Comment (oleg-jukovec) [inline cli/configure/configure.go:199]:
```suggestion
	// Before loading configure file, we'll initialize integrity checking.
```

File: cli/running/script_instance.go [hunk 6]
```diff
@@ -220,6 +244,7 @@ func (inst *scriptInstance) Run(flags RunFlags) error {
 		args = append(args, "-")
 	}
 	args = append(args, flags.RunArgs...)
+
```
Comment (oleg-jukovec) [inline cli/running/script_instance.go:247]: Please, revert the unnecessary change.

File: cli/running/watchdog.go [hunk 7]
```diff
@@ -146,6 +157,27 @@ func (wd *Watchdog) Start() error {
 	return nil
 }
 
+// startIntegrityChecks launches gorountine that performs periodic integrity checks.
+func (wd *Watchdog) startIntegrityChecks() {
+	ticker := time.NewTicker(wd.integrityCheckPeriod)
+
+	go func() {
+		for {
```
Comment (oleg-jukovec) [inline cli/running/watchdog.go:165]: I don't see a normal path exit condition in the loop.

**Review by oleg-jukovec — COMMENTED** [review]
Please, rebase the pull request to a `master` branch + fix the comments below.

File: cli/cluster/file.go
Comment (AlgebraicWolf) [reply cli/cluster/file.go:1]: Fixed
I've rebased the branch on master as well

File: cli/cmd/root.go (hunk 4 above)
Comment (AlgebraicWolf) [reply cli/cmd/root.go]: Fixed

File: cli/configure/configure.go (hunk 5 above)
Comment (AlgebraicWolf) [reply cli/configure/configure.go]: Fixed

File: cli/running/script_instance.go (hunk 6 above)
Comment (AlgebraicWolf) [reply cli/running/script_instance.go]: Fixed

File: cli/running/watchdog.go (hunk 7 above)
Comment (AlgebraicWolf) [reply cli/running/watchdog.go:165]: I've added an exit upon `wd.done`

**Review by oleg-jukovec — APPROVED** [review]
We'll have to continue working later:
https://github.com/tarantool/tt-ee/issues/131
https://github.com/tarantool/tt-ee/issues/132
https://github.com/tarantool/tt-ee/issues/133
